### PR TITLE
Refactor detr 3.x conditional detr

### PR DESCRIPTION
**Pull request by LYMDLUT**
This PR is from https://github.com/open-mmlab/mmdetection/pull/9248 that only without the support of group detr，to merge to main branch

### CODE REVIEW

**Comment by jshilong** [comment]
Please fix the unitest

File: mmdet/models/layers/transformer/utils.py [hunk 1]
```diff
@@ -410,15 +410,49 @@ def forward(self, x: Tensor) -> Tensor:
...
+            Tensor: The output feature, has shape
+                (num_queries, bs, output_dim).
         """
         for i, layer in enumerate(self.layers):
             x = F.relu(layer(x)) if i < self.num_layers - 1 else layer(x)
         return x
 
 
+def gen_sine_embed_for_ref(reference: Tensor):
+    # n_query, bs, _ = pos_tensor.size()
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:424]: please add doc for this function

File: mmdet/models/layers/transformer/conditional_detr_transformer.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+            k_pos = self.kpos_proj(query_pos)
+            v = self.v_proj(query)
+            num_queries, bs, _ = q_content.shape
+            q = q_content if q_pos is None else q_content + q_pos
+            k = k_content if k_pos is None else k_content + k_pos
+            if self.training:
+                q = torch.cat(
+                    q.split(num_queries // self.group_detr, dim=0), dim=1)
+                k = torch.cat(
+                    k.split(num_queries // self.group_detr, dim=0), dim=1)
```
Comment (jshilong) [inline mmdet/models/layers/transformer/conditional_detr_transformer.py:412]: remove all group detr related code

File: mmdet/models/layers/transformer/conditional_detr_transformer.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        super().__init__(init_cfg)
+        self.cross_attn = cross_attn
+        self.keep_query_pos = keep_query_pos
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        self.attn_drop = Dropout(attn_drop)
+        self.proj_drop = Dropout(proj_drop)
+
+        self._init_proj()
+        self.group_detr = group_detr
```
Comment (jshilong) [inline mmdet/models/layers/transformer/conditional_detr_transformer.py:198]: remove this

File: mmdet/models/layers/transformer/conditional_detr_transformer.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        query = self.ffn(query)
+        query = self.norms[2](query)
+
+        return query
+
+
+class ConditionalAttention(BaseModule):
+    """A wrapper of conditional attention, dropout and residual connection."""
+
+    def __init__(self,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/conditional_detr_transformer.py:180]: add doc str for this function

**Comment by jshilong** [comment]
1. Please merge the refacor-detr after #9460 merged
2. remove all group-detr related codes
3. please remove all 91 classes related codes

File: configs/conditional_detr/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+| Backbone |      Model       | Lr schd | Mem (GB) | Inf time (fps) | box AP |                 Config                 |                                                                                                                                    Download                                                                                                                                    |
+| :------: | :--------------: | :-----: | :------: | :------------: | :----: | :------------------------------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+|   R-50   | Conditional DETR |   50e   |   7.9    |                |  41.0  | [config](./detr_r50_8xb2-150e_coco.py) | [model](https://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835-2c4b8974.pth) \| [log](https://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835.log.json) |
```
Comment (hhaAndroid) [inline configs/conditional_detr/README.md:28]: config path is wrong. The weights also need to be updated

File: configs/conditional_detr/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+| :------: | :--------------: | :-----: | :------: | :------------: | :----: | :-----------------------------------------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+|   R-50   | Conditional DETR |   50e   |   7.9    |                |  41.0  | [config](./conditional_detr_r50_8xb2-50e_coco.py) | [model](https://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835-2c4b8974.pth) \| [log](https://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835.log.json) |
```
Comment (jshilong) [inline configs/conditional_detr/README.md:28]: Add a todo here reminding we upload the latest checkpoint

File: mmdet/models/dense_heads/conditional_detr_head.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+    <https://arxiv.org/abs/2108.06152>`_ .
+    """
+
+    def init_weights(self):
+        """Initialize weights of the transformer head."""
+        super().init_weights()
+        # The initialization below for transformer head is very
+        # important as we use Focal_loss for loss_cls
+        if self.loss_cls.use_sigmoid:
```
Comment (jshilong) [inline mmdet/models/dense_heads/conditional_detr_head.py:29]: change to `self.loss_cls.get("use_sigmoid", True): `  to avoid there is no such attribute.

File: mmdet/models/detectors/conditional_detr.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        """
+
+        hidden_states, references = self.decoder(
+            query=query,
+            key=memory,
+            value=memory,
+            query_pos=query_pos,
+            key_pos=memory_pos,
+            key_padding_mask=memory_mask)
+        references = references.transpose(0, 1)
```
Comment (jshilong) [inline mmdet/models/detectors/conditional_detr.py:73]: Can we remove this transpose operation? I believe it is unnecessary

File: mmdet/models/layers/transformer/conditional_detr_transformer.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                as `key_pos`. Defaults to `None`.
+            key_padding_mask (Tensor): ByteTensor with shape (bs, num_keys).
+                Defaults to `None`.
+        Returns:
+            List[Tensor]: forwarded results with shape (num_decoder_layers,
+            bs, num_queries, dim) if `return_intermediate` is True, otherwise
+            with shape (1, bs, num_queries, dim). references with shape
+            (num_decoder_layers, bs, num_queries, 2).
+        """
+        reference_unsigmoid = self.ref_point_head(
```
Comment (jshilong) [inline mmdet/models/layers/transformer/conditional_detr_transformer.py:64]: check the shape comments

File: mmdet/models/layers/transformer/conditional_detr_transformer.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            query_pos)  # [num_queries, batch_size, 2]
+        reference = reference_unsigmoid.sigmoid().transpose(0, 1)
```
Comment (jshilong) [inline mmdet/models/layers/transformer/conditional_detr_transformer.py:66]: too many unnecessary transpose operation for reference, try to make it clear

File: mmdet/models/layers/transformer/utils.py [hunk 11]
```diff
@@ -410,15 +412,67 @@ def forward(self, x: Tensor) -> Tensor:
...
+        scale (float, optional): A scale factor that scales the position
+            embedding. The scale will be used only when `normalize` is True.
+            Defaults to 2*pi.
+    Returns:
+        Tensor: Returned encoded positional tensor.
+    """
+    dim_t = torch.arange(
+        num_feats, dtype=torch.float32, device=pos_tensor.device)
+    dim_t = temperature**(2 * (dim_t // 2) / num_feats)
+    x_embed = pos_tensor[:, :, 0] * scale
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:448]: change all `pos_tensor[:, :, x ]` to `pos_tensor[..., x] `

File: mmdet/models/layers/transformer/utils.py [hunk 12]
```diff
@@ -410,15 +412,67 @@ def forward(self, x: Tensor) -> Tensor:
...
+        pos_w = w_embed[:, :, None] / dim_t
+        pos_w = torch.stack((pos_w[:, :, 0::2].sin(), pos_w[:, :, 1::2].cos()),
+                            dim=3).flatten(2)
+
+        h_embed = pos_tensor[:, :, 3] * scale
+        pos_h = h_embed[:, :, None] / dim_t
+        pos_h = torch.stack((pos_h[:, :, 0::2].sin(), pos_h[:, :, 1::2].cos()),
+                            dim=3).flatten(2)
+
+        pos = torch.cat((pos_y, pos_x, pos_w, pos_h), dim=2)
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:469]: dim=-1

Comment (jshilong) [reply mmdet/models/layers/transformer/utils.py:469]: Changing to [...,x] and dim=-1 here can make it work when pos has different number of dimension

File: mmdet/models/layers/transformer/utils.py [hunk 13]
```diff
@@ -410,15 +412,67 @@ def forward(self, x: Tensor) -> Tensor:
...
-            Tensor: The output feature, has shape (num_query, bs, output_dim).
+            Tensor: The output feature, has shape
+                (num_queries, bs, output_dim).
         """
         for i, layer in enumerate(self.layers):
             x = F.relu(layer(x)) if i < self.num_layers - 1 else layer(x)
         return x
 
 
+def convert_coordinate_to_encoding(pos_tensor: Tensor,
```
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:425]: rename pos_tensor to coord_tensor may be better

**Review by jshilong — APPROVED** [review]
LGTM

File: configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+                dict(type='FocalLossCost', weight=2.0),
+                dict(type='BBoxL1Cost', weight=5.0, box_format='xywh'),
+                dict(type='IoUCost', iou_mode='giou', weight=2.0)
+            ])))
+
+# learning policy
+train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=50, val_interval=1)
+
+param_scheduler = [dict(type='MultiStepLR', end=50, milestones=[40])]
+randomness = dict(seed=42, deterministic=True)
```
Comment (hhaAndroid) [inline configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco.py:43]: is it necessary?

**Review by hhaAndroid — APPROVED** [review]

File: configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco.py (hunk 14 above)
Comment (jshilong) [reply configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco.py:43]: We can remove this seed @LYMDLUT
